src/Featurizers/TfidfVectorizerFeaturizer.h

```diff
@@ -0,0 +1,461 @@
+///
+class TfidfVectorizerTransformer : public StandardTransformer<std::string, TFIDFStruct> {
+public:
+    // ----------------------------------------------------------------------
+    // |
+    // |  Public Types
+    // |
+    // ----------------------------------------------------------------------
+    using BaseType                           = StandardTransformer<std::string, TFIDFStruct>;
+    using IndexMapType                       = std::unordered_map<std::string, std::uint32_t>;
+    using IterRangeType                      = std::tuple<std::string::const_iterator, std::string::const_iterator>;
+
+    // ----------------------------------------------------------------------
+    // |
+    // |  Public Data
+    // |
+    // ----------------------------------------------------------------------
+    IndexMapType const                       Labels;
+    IndexMapType const                       DocumentFreq;
+    std::uint32_t const                      TotalNumsDocuments;
+
+    bool const                               Binary;
+    std::string const                        Norm;
+    bool const                               UseIdf;
+    bool const                               SmoothIdf;
+    bool const                               SublinearTf;
+
+    // ----------------------------------------------------------------------
+    // |
+    // |  Public Methods
+    // |
+    // ----------------------------------------------------------------------
+    TfidfVectorizerTransformer(IndexMapType labels, IndexMapType docufreq, std::uint32_t totalnumdocus, bool binary, 
+                               std::string norm, bool use_idf, bool smooth_idf, bool sublinear_tf);
+    TfidfVectorizerTransformer(Archive &ar);
+
+    ~TfidfVectorizerTransformer(void) override = default;
+
+    void save(Archive &ar) const override;
+
+    bool operator==(TfidfVectorizerTransformer const &other) const;
+
+private:
+    // ----------------------------------------------------------------------
+    // |
+    // |  Private Methods
+    // |
+    // ----------------------------------------------------------------------
+
+    // MSVC has problems when the definition and declaration are separated
+    void execute_impl(typename BaseType::InputType const &input, typename BaseType::CallbackFunction const &callback) override {
+
+        typename std::map<IterRangeType, std::uint32_t, Components::IterRangeComp> TermFreqMap;
+
+        //todo: will use vector<functor> after string header file is done
+        Components::split_temp( 
+            input, 
+            [&TermFreqMap] (std::string::const_iterator & iter_start, std::string::const_iterator & iter_end) {
+                
+                typename std::map<IterRangeType, std::uint32_t, Components::IterRangeComp>::iterator iter_tf(TermFreqMap.find(std::make_tuple(iter_start, iter_end)));
+                
+                if (iter_tf != TermFreqMap.end()) {
+                    ++iter_tf->second;
```

Comment:
> ++iter_tf->second; [](start = 19, length = 19)

Is second represents end iterator? You can't do anything with end iterators except comparision